File: frontend/src/engine/lessonUnits.ts
import type { BossQuestion } from './traceTypes';
import type { LearningAtoms, LearningAtom, LessonUnit } from './lessonTypes';

/**
 * Derives the lesson's unit sequence (the Duolingo-style circles) from the
 * atoms file and the boss-fight questions. Units are derived, never stored.
 *
 * IMPORTANT: this is a mirror of the backend algorithm in
 * backend/.../lesson/LessonUnits.java (pinned there by LessonUnitsTest); any
 * change must be applied to both sides, or circle completion will desync from
 * the server's lesson-completion recompute.
 */

export const PRACTICE_CHUNK = 5;
const MIN_LAST_CHUNK = 3;

function chunk(flat: string[]): string[][] {
  const chunks: string[][] = [];
  for (let i = 0; i < flat.length; i += PRACTICE_CHUNK) {
    chunks.push(flat.slice(i, i + PRACTICE_CHUNK));
  }
  if (chunks.length > 1 && chunks[chunks.length - 1].length < MIN_LAST_CHUNK) {
    const tail = chunks.pop()!;
    chunks[chunks.length - 1].push(...tail);
  }
  return chunks;
}

/** Round-robin the practice of the given atoms (1st of each, then 2nd, ...). */
function roundRobinPractice(atoms: LearningAtom[]): string[] {
  const flat: string[] = [];
  const longest = Math.max(0, ...atoms.map((a) => a.practice?.length ?? 0));
  for (let round = 0; round < longest; round++) {
    for (const atom of atoms) {
      const practice = atom.practice ?? [];
      if (round < practice.length) flat.push(practice[round].id);
    }
  }
  return flat;
}

export function deriveUnits(atoms: LearningAtoms, bossFight: BossQuestion[]): LessonUnit[] {
  const units: LessonUnit[] = [];

  for (const atom of atoms.atoms) {
    if (atom.discovery && atom.discovery.length > 0) {
      units.push({
        id: `d:${atom.id}`,
        kind: 'discovery',
        exerciseIds: atom.discovery.map((e) => e.id),
      });
    }
  }

  const regular = atoms.atoms.filter((a) => !a.capstone);
  const capstones = atoms.atoms.filter((a) => a.capstone);

  chunk(roundRobinPractice(regular)).forEach((ids, i) => {
    units.push({ id: `p${i + 1}`, kind: 'practice', exerciseIds: ids });
  });

  // Capstone synthesis: the capstone atoms' practice, in file order, as a
  // dedicated final block right before the mistakes/boss units.
  const capFlat: string[] = [];
  for (const atom of capstones) {
    for (const e of atom.practice ?? []) capFlat.push(e.id);
  }
  chunk(capFlat).forEach((ids, i) => {
    units.push({ id: `c${i + 1}`, kind: 'capstone', exerciseIds: ids });
  });

  // Single "fix your mistakes" unit; its contents are derived from the
  // learner's wrong answers at render time (see lessonStore).
  units.push({ id: 'mistakes', kind: 'mistakes', exerciseIds: [] });

  for (const q of bossFight) {
    units.push({ id: `b:${q.id}`, kind: 'boss', exerciseIds: [] });
  }
  return units;
}

File: backend/src/main/java/com/interviewlearning/lesson/LessonUnits.java
package com.interviewlearning.lesson;

import com.interviewlearning.lesson.LessonDtos.Atom;
import com.interviewlearning.lesson.LessonDtos.Exercise;
import com.interviewlearning.lesson.LessonDtos.LearningAtoms;
import com.interviewlearning.topics.TopicDtos.BossQuestion;

import java.util.ArrayList;
import java.util.List;

/**
 * Derives the lesson's unit sequence (the Duolingo-style circles) from the
 * atoms file and the boss-fight questions. Units are derived, never stored.
 *
 * <p>IMPORTANT: this algorithm is mirrored on the frontend in
 * {@code frontend/src/engine/lessonUnits.ts}; any change here must be applied
 * there too (the unit test pins the exact output).
 *
 * <ol>
 *   <li>Discovery units: one per atom with a non-empty {@code discovery} list,
 *       in file order; id {@code d:<atomId>}.</li>
 *   <li>Practice units: practice exercises flattened round-robin across atoms
 *       (1st of each atom, then 2nd, ...) so consecutive exercises mix ideas,
 *       chunked into groups of {@link #PRACTICE_CHUNK}; a trailing chunk of
 *       fewer than 3 merges into the previous one; ids {@code p1}, {@code p2}, ...</li>
 *   <li>Boss units: one per boss-fight question, id {@code b:<questionId>}.</li>
 * </ol>
 */
public final class LessonUnits {

    public static final int PRACTICE_CHUNK = 5;
    private static final int MIN_LAST_CHUNK = 3;

    private LessonUnits() {
    }

    /** One derived unit; boss units carry no exercise ids (the question id is in the unit id). */
    public record UnitRef(String unitId, String kind, List<String> exerciseIds) {
    }

    public static List<UnitRef> derive(LearningAtoms atoms, List<BossQuestion> bossFight) {
        List<UnitRef> units = new ArrayList<>();
        List<Atom> all = atoms.atoms() == null ? List.of() : atoms.atoms();

        for (Atom atom : all) {
            List<Exercise> discovery = atom.discovery();
            if (discovery != null && !discovery.isEmpty()) {
                units.add(new UnitRef("d:" + atom.id(), "discovery",
                        discovery.stream().map(Exercise::id).toList()));
            }
        }

        List<String> flat = roundRobinPractice(all);
        List<List<String>> chunks = chunk(flat);
        for (int i = 0; i < chunks.size(); i++) {
            units.add(new UnitRef("p" + (i + 1), "practice", List.copyOf(chunks.get(i))));
        }

        for (BossQuestion q : bossFight) {
            units.add(new UnitRef("b:" + q.id(), "boss", List.of()));
        }
        return units;
    }

    private static List<String> roundRobinPractice(List<Atom> atoms) {
        int longest = atoms.stream()
                .mapToInt(a -> a.practice() == null ? 0 : a.practice().size())
                .max().orElse(0);
        List<String> flat = new ArrayList<>();
        for (int round = 0; round < longest; round++) {
            for (Atom atom : atoms) {
                List<Exercise> practice = atom.practice();
                if (practice != null && round < practice.size()) {
                    flat.add(practice.get(round).id());
                }
            }
        }
        return flat;
    }

    private static List<List<String>> chunk(List<String> flat) {
        List<List<String>> chunks = new ArrayList<>();
        for (int i = 0; i < flat.size(); i += PRACTICE_CHUNK) {
            chunks.add(new ArrayList<>(flat.subList(i, Math.min(i + PRACTICE_CHUNK, flat.size()))));
        }
        if (chunks.size() > 1 && chunks.get(chunks.size() - 1).size() < MIN_LAST_CHUNK) {
            List<String> tail = chunks.remove(chunks.size() - 1);
            chunks.get(chunks.size() - 1).addAll(tail);
        }
        return chunks;
    }
}
